File: node/chart_snapshot.js
#!/usr/bin/env node
// Snapshot a Google Play top chart with StorePulse.
//
// Pulls a top-chart (top_free / top_paid / top_grossing) for a category and
// prints the ranked entries. Run daily to track chart movement over time.
//
// StorePulse — Google Play app metadata, reviews, search rankings & charts as a
// clean, cached JSON API. Stop maintaining your own Play scraper:
// https://rapidapi.com/egosumkira/api/storepulse
//
// Usage:
//   export RAPIDAPI_KEY=...
//   node chart_snapshot.js [collection] [category] [country] [lang]
//   node chart_snapshot.js top_free APPLICATION us
//
// Requires Node 18+ (built-in fetch).

const HOST = "storepulse.p.rapidapi.com";

async function getChart(collection = "top_free", category = "APPLICATION", country = "us", lang = "en") {
  const key = process.env.RAPIDAPI_KEY;
  if (!key) throw new Error("set RAPIDAPI_KEY in your environment");

  const url = new URL(`https://${HOST}/v1/charts`);
  url.search = new URLSearchParams({ collection, category, country, lang }).toString();

  const resp = await fetch(url, {
    headers: { "X-RapidAPI-Key": key, "X-RapidAPI-Host": HOST },
  });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}: ${await resp.text()}`);
  const data = await resp.json();
  return data.charts;
}

async function main() {
  const [collection = "top_free", category = "APPLICATION", country = "us", lang = "en"] =
    process.argv.slice(2);

  console.log(`${collection} / ${category} in ${country}:`);
  for (const entry of await getChart(collection, category, country, lang)) {
    const rank = String(entry.rank).padStart(2);
    console.log(`  #${rank}  ${entry.title}  (${entry.app_id})  ★${entry.score ?? "—"}`);
  }
}

main().catch((err) => {
  console.error(err.message);
  process.exit(1);
});

File: python/chart_snapshot.py
#!/usr/bin/env python3
"""Snapshot a Google Play top chart with StorePulse.

Pulls a top-chart (top_free / top_paid / top_grossing) for a category and prints
the ranked entries. Run daily to track chart movement over time.

StorePulse — Google Play app metadata, reviews, search rankings & charts as a
clean, cached JSON API. Stop maintaining your own Play scraper:
https://rapidapi.com/  (search "StorePulse")

Usage:
    export RAPIDAPI_KEY=...
    python chart_snapshot.py [collection] [category] [country] [lang]
    python chart_snapshot.py top_free APPLICATION us
"""
import os
import sys

import requests

HOST = "storepulse.p.rapidapi.com"


def get_chart(collection="top_free", category="APPLICATION", country="us", lang="en"):
    key = os.environ["RAPIDAPI_KEY"]
    resp = requests.get(
        f"https://{HOST}/v1/charts",
        params={
            "collection": collection,
            "category": category,
            "country": country,
            "lang": lang,
        },
        headers={"X-RapidAPI-Key": key, "X-RapidAPI-Host": HOST},
        timeout=30,
    )
    resp.raise_for_status()
    return resp.json()["charts"]


def main():
    collection = sys.argv[1] if len(sys.argv) > 1 else "top_free"
    category = sys.argv[2] if len(sys.argv) > 2 else "APPLICATION"
    country = sys.argv[3] if len(sys.argv) > 3 else "us"
    lang = sys.argv[4] if len(sys.argv) > 4 else "en"

    print(f"{collection} / {category} in {country}:")
    for entry in get_chart(collection, category, country, lang):
        score = entry.get("score", "—")
        print(f"  #{entry['rank']:>2}  {entry['title']}  ({entry['app_id']})  ★{score}")


if __name__ == "__main__":
    main()
